### PR TITLE
Spark 3.4: Add RewritePositionDeleteFilesProcedure

### PR DESCRIPTION
**Pull request by szehon-ho**
Adds a spark procedure: rewrite_position_delete_files for #7389

### CODE REVIEW

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                    + "table => '%s',"
+                    + "options => map("
+                    + "'foo', 'bar'))",
+                catalogName, tableIdent));
+  }
+
+  private Map<String, String> snapshotSummary() {
+    return validationCatalog.loadTable(tableIdent).currentSnapshot().summary();
+  }
+}
```
Comment (ajantha-bhat) [inline spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java:152]: can we also add a test case with dangling deletes?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+          Result result = action.execute();
+
+          return toOutputRows(result);
+        });
+  }
+
+  private InternalRow[] toOutputRows(Result result) {
+    int rewrittenDataFilesCount = result.rewrittenDeleteFilesCount();
+    long rewrittenBytesCount = result.rewrittenBytesCount();
+    int addedDataFilesCount = result.addedDeleteFilesCount();
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java]: Thanks for the catch

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedures.java [hunk 3]
```diff
@@ -54,6 +54,7 @@ private static Map<String, Supplier<ProcedureBuilder>> initProcedureBuilders() {
     mapBuilder.put("register_table", RegisterTableProcedure::builder);
     mapBuilder.put("publish_changes", PublishChangesProcedure::builder);
     mapBuilder.put("create_changelog_view", CreateChangelogViewProcedure::builder);
+    mapBuilder.put("rewrite_position_delete_files", RewritePositionDeleteFilesProcedure::builder);
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedures.java:57]: What does everyone think?  I use RewritePositionDeletes and PositionDeletes in the code a lot for shortness, but was not sure as the procedure names all indicate files

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        "Should delete 2 delete files and add 1",
+        ImmutableList.of(
+            row(
+                2,
+                1,
+                Long.valueOf(snapshotSummary.get(REMOVED_FILE_SIZE_PROP)),
+                Long.valueOf(snapshotSummary.get(ADDED_FILE_SIZE_PROP)))),
+        output);
+
+    Assert.assertEquals(1, TestHelpers.deleteFiles(table).size());
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java:99]: I wanted this test to be about the procedure code, as I already have added coverage of the Action itself in TestRewritePositionDeleteFilesAction

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java (hunk 1 above)
Comment (szehon-ho) [reply spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java:152]: Same, added already tests  of dangling deletes on : TestRewritePositionDeleteFilesAction, and was thinking this to be a quicker test just to validate the procedure code only

**Review by ajantha-bhat — APPROVED** [review]

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedures.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedures.java:57]: Normally I'd go for shorter names, but I see here all the keys in the map already have `files` suffix so I think it's fine as it is currently

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    ProcedureInput input = new ProcedureInput(spark(), tableCatalog(), PARAMETERS, args);
+
+    Identifier tableIdent = input.ident(TABLE_PARAM);
+
+    Map<String, String> options = input.asStringMap(OPTIONS_PARAM, ImmutableMap.of());
```
Comment (amogh-jahagirdar) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java:85]: Nit: I think the new lines at 82 and 84 here are a bit unnecessary (think the one before modifyIcebergTable call makes sense though)

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  @After
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+  }
+
+  @Test
+  public void testExpireDeleteFilesAll() throws Exception {
+    createTable();
```
Comment (amogh-jahagirdar) [inline spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java:71]: Can we move the `createTable()` to a "Before"?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.connector.iceberg.catalog.ProcedureParameter;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+public class RewritePositionDeleteFilesProcedure extends BaseProcedure {
```
Comment (amogh-jahagirdar) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java:34]: I know it's somewhat self explanatory but would we add a class level javadoc similar to `RewriteDataFilesProcedure` linking to the action?

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java (hunk 6 above)
Comment (szehon-ho) [reply spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java:71]: Is it ok to keep manually for now?  I plan to add filter at some point and so will probably add a test for another kind of table (like partitioned table).  Usually in my experience it becomes awkward to support test of different kind of table, if the before is already creating one kind of table.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java (hunk 7 above)
Comment (szehon-ho) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java:40]: Added

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+  private static final ProcedureParameter TABLE_PARAM =
+      ProcedureParameter.required("table", DataTypes.StringType);
+  private static final ProcedureParameter OPTIONS_PARAM =
+      ProcedureParameter.optional("options", STRING_MAP);
+
+  private static final ProcedureParameter[] PARAMETERS =
+      new ProcedureParameter[] {TABLE_PARAM, OPTIONS_PARAM};
+
+  // counts are not nullable since the action result is never null
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java]: Removed

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java (hunk 5 above)
Comment (szehon-ho) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java]: Removed

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedures.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (szehon-ho) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedures.java:57]: Yea thats the primary reason why i kept 'files' though i do think it is long, @aokolnychyi what do you think?

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java (hunk 6 above)
Comment (amogh-jahagirdar) [reply spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewritePositionDeleteFilesProcedure.java:71]: Sure that makes sense!

**Review by amogh-jahagirdar — APPROVED** [review]
Looks good to me, thanks @szehon-ho !

**Comment by aokolnychyi** [comment]
Let me take a look as well.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedures.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/SparkProcedures.java:57]: No strong opinion. I also like shorter names but it seems we use `files` suffix in other procedures so I'll be inclined to keep the name as is.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        table -> {
+          RewritePositionDeleteFiles action =
+              actions().rewritePositionDeletes(table).options(options);
+          Result result = action.execute();
+
+          return toOutputRows(result);
+        });
+  }
+
+  private InternalRow[] toOutputRows(Result result) {
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java:101]: Optional: Another way to write it.

```
private InternalRow toOutputRow(Result result) {
  return newInternalRow(
      result.rewrittenDeleteFilesCount(),
      result.rewrittenBytesCount(),
      result.addedDeleteFilesCount(),
      result.addedBytesCount());
}
```

```
...
return new InternalRow[] {toOutputRow(result)};
```

Up to you, though.

**Review by aokolnychyi — APPROVED** [review]

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    ProcedureInput input = new ProcedureInput(spark(), tableCatalog(), PARAMETERS, args);
+    Identifier tableIdent = input.ident(TABLE_PARAM);
+    Map<String, String> options = input.asStringMap(OPTIONS_PARAM, ImmutableMap.of());
+
+    return modifyIcebergTable(
+        tableIdent,
+        table -> {
+          RewritePositionDeleteFiles action =
```
Comment (szehon-ho) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java:93]: Just noticed I can get remove this as well and just chain the action call

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java (hunk 9 above)
Comment (szehon-ho) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/procedures/RewritePositionDeleteFilesProcedure.java]: Yep, figured out your argument is in the wrong order, but yep, done.

**Comment by aokolnychyi** [comment]
Thanks everyone!

**Pull request merged**